### PR TITLE
Caom2.4

### PR DESCRIPTION
**Pull request by andamian**
S2443

### CODE REVIEW

File: caom2/caom2/tests/caom_test_instances.py [hunk 1]
```diff
@@ -562,6 +576,14 @@ def get_polarization_wcs(self):
...
         return pol
 
+    def get_custom_wcs(self):
+        axis = wcs.Axis('Foo')
+        axis1d = wcs.CoordAxis1D(axis)
+        axis1d.function = wcs.CoordFunction1D(int(5), 1.0,
+                                              wcs.RefCoord(1.0, 1.0))
+        custom = chunk.CustomWCS(axis1d)
+        return custom
+
```
Comment (SharonGoliath) [inline caom2/caom2/tests/caom_test_instances.py]: Do there need to be test instances for the ReadGroup fields?

File: caom2/setup.cfg [hunk 2]
```diff
@@ -37,7 +37,7 @@ license = AGPLv3
 url = http://www.cadc-ccda.hia-iha.nrc-cnrc.gc.ca/caom2
 edit_on_github = False
 github_project = opencadc/caom2tools
-install_requires = future lxml aenum six
+install_requires = future lxml aenum six deprecated
 # version should be PEP386 compatible (http://www.python.org/dev/peps/pep-0386)
 version = 2.3.8.4
```
Comment (SharonGoliath) [inline caom2/setup.cfg]: Will the new version be 2.4?

File: caom2/caom2/caom_util.py [hunk 3]
```diff
@@ -116,16 +115,15 @@ def validate_path_component(caller, name, test):
...
-            '/' not in test and
-            '||' not in test and
-            '%' not in test), (
-        caller.__class__.__name__ + ": invalid " + name +
-        ": may not contain space ( ), slash (/), escape (\\), or percent (%)")
+    if ' ' in test or '/' in test or '||' in test or '%' in test:
+        raise \
+            ValueError(caller.__class__.__name__ + ": invalid " + name +
+                       ": may not contain space ( ), slash (/), escape (\\), "
+                       "or percent (%)")
```
Comment (SharonGoliath) [inline caom2/caom2/caom_util.py]: Not sure how to comment on code that is not changed, so I'm putting this comment here: there are still 'assert' statements in common.py, diff.py, obs_reader_writer.py, observation.py, plane.py and shape.py.

File: caom2/caom2/chunk.py [hunk 4]
```diff
@@ -275,10 +282,21 @@ def time_axis(self):
 
     @time_axis.setter
     def time_axis(self, value):
-        caom_util.type_check(value, int_32, 'polarization_axis')
+        caom_util.type_check(value, int_32, 'time_axis')
```
Comment (SharonGoliath) [inline caom2/caom2/chunk.py]: Remove the commented out code?

File: caom2/setup.cfg (hunk 2 above)
Comment (andamian) [reply caom2/setup.cfg]: Yes. 2.4 or maybe a pre-release first.

File: caom2/caom2/tests/caom_test_instances.py (hunk 1 above)
Comment (andamian) [reply caom2/caom2/tests/caom_test_instances.py]: Added. Bet it doesn't hurt for code coverage...

File: caom2/caom2/tests/caom_test_instances.py [hunk 5]
```diff
@@ -313,12 +313,11 @@ def get_planes(self):
                     _plane.position = self.get_poly_position()
                 if s == 'circle':
                     _plane.position = self.get_circle_position()
-                if self.caom_version >= 22:
-                    _plane.energy = self.get_energy()
-                    _plane.time = self.get_time()
-                    _plane.polarization = self.get_polarization()
-                if self.caom_version == 21:
-                    _plane.quality = self.get_quality()
```
[the remark this answers is not in the document]
Comment (andamian) [reply caom2/caom2/tests/caom_test_instances.py]: Removed everywhere in this file.

File: caom2/caom2/plane.py [hunk 6]
```diff
@@ -265,13 +196,35 @@ class PolarizationState(Enum):
...
 
 class Quality(Enum):
     """
     JUNK: junk
     """
-    JUNK = "junk"
+    JUNK = VocabularyTerm(_CAOM_VOCAB_NS, "junk", True).get_value()
+
+
+class Observable():
```
[the remark this answers is not in the document]
Comment (andamian) [reply caom2/caom2/plane.py]: Added. Thanks

File: caom2/caom2/observation.py [hunk 7]
```diff
@@ -215,6 +222,7 @@ def __init__(self,
         self.target_position = target_position
         self.requirements = requirements
         self.meta_release = meta_release
+        self.meta_read_groups = meta_read_groups
```
[the remark this answers is not in the document]
Comment (andamian) [reply caom2/caom2/observation.py]: Good point. Fixed.

File: caom2/caom2/obs_reader_writer.py [hunk 8]
```diff
@@ -1966,22 +2103,38 @@ def _add_position_element(self, position, parent):
         self._add_shape_element("bounds", position.bounds, element)
         self._add_dimension2d_element("dimension", position.dimension, element)
         self._add_element("resolution", position.resolution, element)
+        self._add_interval_element("resolutionBounds",
```
[the remark this answers is not in the document]
Comment (andamian) [reply caom2/caom2/obs_reader_writer.py]: Fixed.

File: caom2/caom2/obs_reader_writer.py [hunk 9]
```diff
@@ -1966,22 +2103,38 @@ def _add_position_element(self, position, parent):
...
 
     def _add_energy_element(self, energy, parent):
         if energy is None:
             return
-
         element = self._get_caom_element("energy", parent)
         self._add_interval_element("bounds", energy.bounds, element)
         self._add_element("dimension", energy.dimension, element)
         self._add_element("resolvingPower", energy.resolving_power, element)
+        self._add_interval_element("resolvingPowerBounds",
```
[the remark this answers is not in the document]
Comment (andamian) [reply caom2/caom2/obs_reader_writer.py]: And done

File: caom2/caom2/obs_reader_writer.py [hunk 10]
```diff
@@ -1992,18 +2145,26 @@ def _add_energy_element(self, energy, parent):
     def _add_time_element(self, time, parent):
         if time is None:
             return
-
         element = self._get_caom_element("time", parent)
         self._add_interval_element("bounds", time.bounds, element)
         self._add_element("dimension", time.dimension, element)
         self._add_element("resolution", time.resolution, element)
+        self._add_interval_element("resolutionBounds",
```
[the remark this answers is not in the document]
Comment (andamian) [reply caom2/caom2/obs_reader_writer.py]: Done as well.

File: caom2/caom2/common.py [hunk 11]
```diff
@@ -98,6 +106,42 @@ def get_current_ivoa_time():
                     now.second, int(str(now.microsecond)[:-3] + '000'))
 
 
+class OrderedEnum(Enum):
```
[the remark this answers is not in the document]
Comment (andamian) [reply caom2/caom2/common.py]: It's internal - not part of the CAOM2 model.

File: caom2/caom2/chunk.py (hunk 4 above)
Comment (andamian) [reply caom2/caom2/chunk.py]: Done

File: caom2/caom2/caom_util.py [hunk 12]
```diff
@@ -358,6 +330,37 @@ def __contains__(self, item):
...
+    def __init__(self, scheme=None, *args):
+        """
+        Arguments:
+        scheme: Enforce a particular scheme, ex 'ivo'
+        *args : values to be stored in the set
+
+        AssertionError will be raised if the types of the arguments
+        are not found valid URIs
+        """
+        self.scheme = scheme
```
[the remark this answers is not in the document]
Comment (andamian) [reply caom2/caom2/caom_util.py]: Scheme is allowed to be None

File: caom2/caom2/caom_util.py (hunk 3 above)
Comment (andamian) [reply caom2/caom2/caom_util.py]: Phew. `caom2` is now `assert` free!

File: caom2/caom2/artifact.py [hunk 13]
```diff
@@ -245,6 +251,36 @@ def content_checksum(self, value):
...
+    @property
+    def content_read_groups(self):
+        return self._content_read_groups
+
+    @content_read_groups.setter
+    def content_read_groups(self, value):
+        """
+            value is a caom_util.URISet
+        """
+        if value is None:
```
[the remark this answers is not in the document]
Comment (andamian) [reply caom2/caom2/artifact.py]: Updated parts setter instead.

File: caom2/caom2/artifact.py [hunk 14]
```diff
@@ -245,6 +251,36 @@ def content_checksum(self, value):
...
+    @property
+    def content_release(self):
+        return self._content_release
+
+    @content_release.setter
+    def content_release(self, value):
+        if value:
+            caom_util.type_check(value, datetime, 'content_release')
+            caom_util.value_check(value,
+                                  datetime(1800, 1, 1, 0, 0, 0),
```
[the remark this answers is not in the document]
Comment (andamian) [reply caom2/caom2/artifact.py]: Good point. Done

**Comment by andamian** [comment]
> Does the wcs validator need to include custom?

That's a good question. I don't know how can a custom wcs be validated.

**Pull request merged**